src/file-manager/fm-desktop-wayland-bg-dialog.c

```diff
@@ -0,0 +1,454 @@
+   License, or (at your option) any later version.
+
+   The Mate Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Library General Public License for more details.
+
+   You should have received a copy of the GNU Library General Public
+   License along with the Mate Library; see the file COPYING.LIB.  If not,
+   write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+
+   Author: <lukefromdc@hushmail.com>
+*/
+
+#include <config.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <X11/Xatom.h>
+#include <gtk/gtk.h>
+#include <gdk/gdkx.h>
+#include <glib/gi18n.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#include <eel/eel-glib-extensions.h>
+#include <eel/eel-gtk-extensions.h>
+#include <eel/eel-vfs-extensions.h>
+
+#include <libcaja-private/caja-directory-background.h>
+#include <libcaja-private/caja-global-preferences.h>
+
+#include "fm-desktop-icon-view.h"
+#include "fm-desktop-wayland-bg-dialog.h"
+
+
+#ifdef HAVE_WAYLAND
+
+GSettings   *background_settings;
+
+static void
+update_preview (gboolean starting, GtkWidget *box, const gchar *filename,
+               const gchar *shading_type, const gchar *primary_color_str, const gchar  *secondary_color_str)
+{
+    static GtkWidget *preview_image;
+    static GtkCssProvider *provider;
+    gchar *css;
+    GString *string;
+    static GdkRectangle geometry = {0};
+
+    /* setup the preview only once*/
+    if (starting == TRUE)
+    {
+        static GtkWidget *preview;
+
+        /*Get the size and shape of the desktop*/
+        GdkDisplay *display = gdk_screen_get_display (gdk_screen_get_default());
+        GdkMonitor *monitor = gdk_display_get_monitor (display, 0);
+        gdk_monitor_get_geometry (monitor, &geometry);
+
+        preview = gtk_box_new (GTK_ORIENTATION_VERTICAL, 0);
+        gtk_widget_set_size_request (preview, geometry.width / 5, geometry.height / 5);
+        gtk_widget_set_name (GTK_WIDGET (preview), "caja-wayland-bg-preview");
+
+        preview_image = gtk_image_new ();
+        provider = gtk_css_provider_new ();
+        gtk_style_context_add_provider (gtk_widget_get_style_context (preview),
+                                        GTK_STYLE_PROVIDER (provider),
+                                        GTK_STYLE_PROVIDER_PRIORITY_APPLICATION);
+        gtk_box_pack_start (GTK_BOX (preview), preview_image, FALSE, FALSE, 0);
+        gtk_box_pack_start (GTK_BOX (box), preview, FALSE, FALSE, 0);
+    }
+
+    /* No image filename means we are previewing a color or gradient background*/
+    if ((!filename) || (strcmp (filename, "") == 0) || (strcmp (filename, " ") == 0))
+
+    {
+        gtk_image_clear (GTK_IMAGE(preview_image));
```

Comment:
This is very dangerous because the ```preview_image``` may not have been initialized

Reply:
We have as line 80
        preview_image = gtk_image_new ();
in the     if (starting == TRUE)
block of `update_preview`
Since we don't want to make a new image each time, I will add a check to clear the image only if it exists.